dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/BackwardsReachingDefinitionsAnalysis.java

```diff
@@ -0,0 +1,384 @@
+ * this implementation analyzes the program from back to front,
+ * assigning each program point the set of readers,
+ * who may still require a register initialization.
+ * This means that it does not collect definitions for unused registers.
+ * Especially, it does not collect last writers for all registers.
+ * <p>
+ * This analysis is control-flow-sensitive;
+ * that is, {@link Label} splits the information among the jumps to it
+ * and {@link CondJump} merges the information.
+ * <p>
+ * This analysis supports loops;
+ * that is, backward jumps cause re-evaluation of the loop body until convergence.
+ * This results in a squared worst-case time complexity in terms of events being processed.
+ */
+public class BackwardsReachingDefinitionsAnalysis implements ReachingDefinitionsAnalysis {
+
+    private final Map<RegReader, ReaderInfo> readerMap = new HashMap<>();
+    private final Map<RegWriter, Readers> writerMap = new HashMap<>();
+
+    @Override
+    public Writers getWriters(RegReader reader) {
+        Preconditions.checkNotNull(reader, "reader is null");
+        final ReaderInfo result = readerMap.get(reader);
+        Preconditions.checkArgument(result != null, "reader %s has not been analyzed.", reader);
+        return result;
+    }
+
+    @Override
+    public Writers getFinalWriters() {
+        final ReaderInfo result = readerMap.get(null);
+        Preconditions.checkState(result != null, "final state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Lists all potential users of the result of an event.
+     * @param writer Event of interest to fetch information about.
+     * @throws IllegalArgumentException If {@code writer} was not inside the program when it was analyzed.
+     */
+    public Readers getReaders(RegWriter writer) {
+        Preconditions.checkNotNull(writer, "writer is null");
+        final Readers result = writerMap.get(writer);
+        Preconditions.checkArgument(result != null, "writer %s has not been analyzed.", writer);
+        return result;
+    }
+
+    /**
+     * Lists all potential users of uninitialized registers.
+     */
+    public Readers getInitialReaders() {
+        final Readers result = writerMap.get(null);
+        Preconditions.checkState(result != null, "initial state has not been analyzed.");
+        return result;
+    }
+
+    /**
+     * Analyzes an entire function definition.
+     * @param function Defined function to be analyzed.  Its parameters are marked as initialized.
+     */
+    public static BackwardsReachingDefinitionsAnalysis forFunction(Function function) {
+        final var analysis = new BackwardsReachingDefinitionsAnalysis();
+        final Set<Register> finalRegisters = new HashSet<>();
+        analysis.initialize(function, finalRegisters);
+        analysis.run(function, finalRegisters);
+        analysis.postProcess();
+        return analysis;
+    }
+
+    /**
+     * Analyzes an entire set of threads.
```

Comment:
"an entire set of threads." -> "the entire program (after thread creation)" ... you are not passing any threads as parameter